src/shared/label-util.c

```diff
@@ -115,3 +116,30 @@ int btrfs_subvol_make_label(const char *path) {
 
         return mac_smack_fix(path, 0);
 }
+
+static int label_pre(int dir_fd, const char *path, mode_t mode) {
+        return mac_selinux_create_file_prepare_at(dir_fd, path, mode);
+}
+
+static int label_post(int dir_fd, const char *path) {
+        mac_selinux_create_file_clear();
+        return mac_smack_fix_full(dir_fd, path, NULL, 0);
+}
+
+static const LabelOps *label_ops = &(LabelOps) {
+        .pre = label_pre,
+        .post = label_post,
+};
+
+int mac_init(void) {
+        int r;
+
+        if (!mac_selinux_use() && !mac_smack_use())
+                return 0;
+
+        r = mac_selinux_init();
+        if (r < 0)
+                return r;
+
+        return label_ops_set(label_ops);
+}
```

Comment:
so i am still puzzled about the implementation here. I would have assumed that 

mac-selinux.c implement one LabelOps structure, and that mac_selinux_init() installs it via label_ops_set() if it detects that selinux is actually is used. And similar, mac-smack.c implements a second LabelOps structure, and that mac_smack_init() installs it via label_ops_set() if it detects that smack is actually is used.

Yes, this would mean you cannot have both selinux and smack enabled in the kernel, but the thing is: you really cannot anyway, the kernel's label APIs only allow returning one label per object right now, and this is going to be very hard to change.

(and if we really wanted to support that we could implement third LabelOps structure if we wanted to that calls into both subsystems, and is only installed if really both are enabled).